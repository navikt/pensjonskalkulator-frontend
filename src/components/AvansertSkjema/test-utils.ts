export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
