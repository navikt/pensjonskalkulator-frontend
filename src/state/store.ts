import {
  combineReducers,
  configureStore,
  createListenerMiddleware,
  ListenerEffectAPI,
  TypedStartListening,
  TypedAddListener,
  PreloadedState,
} from '@reduxjs/toolkit'

import { apiSlice } from './api/apiSlice'
import { createUttaksalderListener } from './listeners/uttaksalderListener'
import userInputReducer from './userInput/userInputReducer'

const listenerMiddleware = createListenerMiddleware()

export const rootReducer = combineReducers({
  [apiSlice.reducerPath]: apiSlice.reducer,
  userInput: userInputReducer,
})

export function setupStore(
  preloadedState?: PreloadedState<RootState>,
  isDev = false
) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware(
        isDev ? { immutableCheck: false, serializableCheck: false } : {}
      )
        .concat(apiSlice.middleware)
        .prepend(listenerMiddleware.middleware),
  })
}

export const store = setupStore()

createUttaksalderListener(
  listenerMiddleware.startListening as AppStartListening
)

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']

export type AppListenerEffectAPI = ListenerEffectAPI<RootState, AppDispatch>
export type AppStartListening = TypedStartListening<RootState, AppDispatch>
export type AppAddListener = TypedAddListener<RootState, AppDispatch>
