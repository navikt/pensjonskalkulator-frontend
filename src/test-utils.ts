import renderer from 'react-test-renderer'

import { render } from '@testing-library/react'

export function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => children,
    ...options,
  })

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
