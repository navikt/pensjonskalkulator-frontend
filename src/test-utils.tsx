import React, { PropsWithChildren } from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import { PreloadedState, createListenerMiddleware } from '@reduxjs/toolkit'
import { render, RenderOptions } from '@testing-library/react'

import { getTranslation_test } from '@/utils/__tests__/test-translations'

import { createUttaksalderListener } from './state/listeners/uttaksalderListener'
import {
  setupStore,
  RootState,
  AppStore,
  AppStartListening,
} from './state/store'
import { getTranslation_nb } from './translations/nb'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  hasRouter?: boolean
}

export const swallowErrors = (testFn: () => void) => {
  const error = console.error
  console.error = () => {}
  testFn()
  console.error = error
}

export const swallowErrorsAsync = async (testFn: () => Promise<void>) => {
  const cache = console.error
  console.error = () => {}
  await testFn()
  console.error = cache
}

function generateMockedTranslations() {
  const nbTranslations: Record<string, string> = getTranslation_nb()
  const testTranslations: Record<string, string> = getTranslation_test()
  const translationsInput = { ...nbTranslations, ...testTranslations }
  const translations: Record<string, string> = {}

  for (const key in translationsInput) {
    if (
      /<(?=.*? .*?\/ ?>|br|hr|input|!--|wbr)[a-z]+.*?>|<([a-z]+).*?<\/\1>|{[^}]+}/i.test(
        translationsInput[key]
      )
    ) {
      // for html or chunks keys: results in 'my_key' : 'my_key with some <html> or {chunk}'
      translations[key] = translationsInput[key]
    } else {
      // results in 'my_key' : 'my_key'
      translations[key] = key
    }
  }
  return translations
}

// Return an object with the store and all of RTL's query functions
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState, true),
    hasRouter = true,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <IntlProvider locale="nb" messages={generateMockedTranslations()}>
          {hasRouter ? <MemoryRouter>{children}</MemoryRouter> : children}
        </IntlProvider>
      </Provider>
    )
  }

  const listenerMiddleware = createListenerMiddleware()
  createUttaksalderListener(
    listenerMiddleware.startListening as AppStartListening
  )

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'

export { renderWithProviders as render }
