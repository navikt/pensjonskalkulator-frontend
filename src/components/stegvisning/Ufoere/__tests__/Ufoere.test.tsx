import { describe, it, vi } from 'vitest'

import { Ufoere } from '..'
import { RootState } from '@/state/store'
import { screen, render, waitFor, userEvent } from '@/test-utils'

describe('stegvisning - Ufoere', () => {
  const onCancelMock = vi.fn()
  const onPreviousMock = vi.fn()
  const onNextMock = vi.fn()

  it('rendrer slik den skal', async () => {
    const user = userEvent.setup()
    const result = render(
      <Ufoere
        isLastStep={false}
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />
    )
    expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(
      'stegvisning.ufoere.title'
    )

    expect(await screen.findByText('stegvisning.ufoere.info')).toBeVisible()

    await user.click(screen.getByText('stegvisning.ufoere.readmore_1.title'))
    expect(
      await screen.findByText(
        'For å ha rett til AFP, kan du ikke ha fått utbetalt uføretrygd',
        { exact: false }
      )
    ).toBeVisible()
    await user.click(screen.getByText('stegvisning.ufoere.readmore_2.title'))
    expect(
      await screen.findByText('Hvis du jobber i privat sektor,', {
        exact: false,
      })
    ).toBeVisible()
    expect(await screen.findByText('stegvisning.ufoere.ingress')).toBeVisible()
    expect(result.asFragment()).toMatchSnapshot()
  })

  it('kaller onNext når det er siste steg og at brukeren klikker på Beregn', async () => {
    const user = userEvent.setup()
    render(
      <Ufoere
        isLastStep={true}
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />
    )
    expect(screen.queryByText('stegvisning.neste')).not.toBeInTheDocument()
    await user.click(screen.getByText('stegvisning.beregn'))

    waitFor(() => {
      expect(onNextMock).toHaveBeenCalled()
    })
  })

  it('kaller onPrevious når brukeren klikker på Tilbake', async () => {
    const user = userEvent.setup()
    render(
      <Ufoere
        isLastStep={false}
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />,
      {
        preloadedState: {
          userInput: { samtykke: true, afp: 'ja_privat' },
        } as RootState,
      }
    )

    await user.click(screen.getByText('stegvisning.tilbake'))
    expect(onPreviousMock).toHaveBeenCalled()
  })

  it('kaller onCancel når brukeren klikker på Avbryt', async () => {
    const user = userEvent.setup()
    render(
      <Ufoere
        isLastStep={false}
        onCancel={onCancelMock}
        onPrevious={onPreviousMock}
        onNext={onNextMock}
      />
    )

    await user.click(screen.getByText('stegvisning.avbryt'))
    expect(onCancelMock).toHaveBeenCalled()
  })
})
