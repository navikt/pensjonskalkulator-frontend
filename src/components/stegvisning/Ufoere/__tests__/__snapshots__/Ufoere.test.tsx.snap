// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stegvisning - Ufoere > rendrer slik den skal 1`] = `
<DocumentFragment>
  <section
    class="_section_67454f _section__largePadding_67454f _section__marginBotton_67454f"
  >
    <form>
      <h2
        class="navds-heading navds-heading--medium navds-typo--spacing"
      >
        stegvisning.ufoere.title
      </h2>
      <div
        aria-live="polite"
        class="_alert_10fc87 navds-alert navds-alert--info navds-alert--medium"
        data-testid="ufoere-info"
      >
        <svg
          aria-labelledby="title-r0"
          class="navds-alert__icon"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="title-r0"
          >
            Informasjon
          </title>
          <path
            clip-rule="evenodd"
            d="M3.25 4A.75.75 0 0 1 4 3.25h16a.75.75 0 0 1 .75.75v16a.75.75 0 0 1-.75.75H4a.75.75 0 0 1-.75-.75zM11 7.75a1 1 0 1 1 2 0 1 1 0 0 1-2 0m-1.25 3a.75.75 0 0 1 .75-.75H12a.75.75 0 0 1 .75.75v4.75h.75a.75.75 0 0 1 0 1.5h-3a.75.75 0 0 1 0-1.5h.75v-4h-.75a.75.75 0 0 1-.75-.75"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <div
          class="navds-alert__wrapper navds-alert__wrapper--maxwidth navds-body-long navds-body-long--medium"
        >
          Før du fyller 62 år må du velge mellom å få AFP eller å beholde uføretrygden. 
          <br />
          <br />
           AFP og uføretrygd kan ikke kombineres. Hvis du ikke gir oss beskjed, mister du retten til AFP (men beholder uføretrygden).
        </div>
      </div>
      <div
        class="navds-read-more navds-read-more--medium _readmore1_10fc87 navds-read-more--open"
      >
        <button
          aria-expanded="true"
          class="navds-read-more__button navds-body-short"
          data-testid="readmore"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="navds-read-more__expand-icon"
            fill="none"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M5.97 9.47a.75.75 0 0 1 1.06 0L12 14.44l4.97-4.97a.75.75 0 1 1 1.06 1.06l-5.5 5.5a.75.75 0 0 1-1.06 0l-5.5-5.5a.75.75 0 0 1 0-1.06"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <span>
            stegvisning.ufoere.readmore_1.title
          </span>
        </button>
        <div
          aria-hidden="false"
          class="navds-read-more__content navds-body-long navds-body-long--medium"
        >
          For å ha rett til AFP, kan du ikke ha fått utbetalt uføretrygd fra Nav etter den måneden du fyller 62 år. Det gjelder uansett om du har mottatt hel eller gradert uføretrygd, hvor lenge du har hatt uføretrygd og hvor mye du har fått utbetalt i uføretrygd.
          <br />
          <br />
          Hvis du er under 62 år, må du altså si fra deg uføretrygden innen utgangen av måneden du fyller 62 år for å få utbetalt AFP. Husk at alle de andre vilkårene for å ha rett til AFP også må være oppfylt.
        </div>
      </div>
      <p
        class="_paragraph_10fc87 navds-body-long navds-body-long--large"
        data-testid="ufoere-ingress"
      >
        Du kan få hjelp til å vurdere alternativene dine. Kontakt tjenestepensjonsordningen din hvis du jobber i offentlig sektor. 
        <a
          class="navds-link navds-link--action navds-link--inline-text"
          href="https://www.nav.no/planlegger-pensjon#noe-du-ikke-finner-svaret-p-her"
          target="_blank"
        >
          Kontakt Nav
          <svg
            aria-labelledby="title-r2"
            fill="none"
            focusable="false"
            height="1.25rem"
            role="img"
            viewBox="0 0 24 24"
            width="1.25rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title
              id="title-r2"
            >
              application.global.external_link
            </title>
            <path
              clip-rule="evenodd"
              d="M20.532 3.471A.75.75 0 0 1 20.75 4v7.5a.75.75 0 0 1-1.5 0V5.81l-8.72 8.72a.75.75 0 1 1-1.06-1.06l8.72-8.72H12.5a.75.75 0 0 1 0-1.5H20c.206 0 .393.083.529.218l.001.002zM4.75 9A.25.25 0 0 1 5 8.75h7a.75.75 0 0 0 0-1.5H5A1.75 1.75 0 0 0 3.25 9v10c0 .966.784 1.75 1.75 1.75h10A1.75 1.75 0 0 0 16.75 19v-7a.75.75 0 0 0-1.5 0v7a.25.25 0 0 1-.25.25H5a.25.25 0 0 1-.25-.25z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </a>
         hvis du jobber i privat sektor. 
        <br />
        <br />
         Kalkulatoren beregner ikke AFP for deg som får uføretrygd.
        <br />
        <br />
         Gå videre for å se alderspensjon fra Nav og pensjonsavtaler i privat sektor.
      </p>
      <button
        class="_button_10fc87 navds-button navds-button--primary navds-button--medium"
        type="submit"
      >
        <span
          class="navds-label"
        >
          stegvisning.neste
        </span>
      </button>
      <button
        class="_button_10fc87 navds-button navds-button--secondary navds-button--medium"
        type="button"
      >
        <span
          class="navds-label"
        >
          stegvisning.tilbake
        </span>
      </button>
      <button
        class="_button_10fc87 navds-button navds-button--tertiary navds-button--medium"
        type="button"
      >
        <span
          class="navds-label"
        >
          stegvisning.avbryt
        </span>
      </button>
    </form>
  </section>
</DocumentFragment>
`;
